opt: Store root expression directly in Memo

andy-kimball opened the pull request:
Currently, the memo's root expression is maintained outside the memo.
There are a few cases (like printing the Memo) where having it
available in the memo would be convenient. This change sets the root
expression (defined as the root of the lowest cost expression tree)
just after optimization is complete. The memo formatting code now
uses that information directly. Future PR's will make further usage
of the root.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24712)
<!-- Reviewable:end -->

andy-kimball commented:
@justinj, I know this will probably conflict with your change. Once you've merged I can reconcile any conflicts.

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/opt/memo/memo.go, line 129 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-L9tuitw0bBiM4cWkbBi:-L9tuitw0bBiM4cWkbBj:b-nakdt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7b586f08aeb888382ac0fa2579aa0cb86258d1f/pkg/sql/opt/memo/memo.go#L129)):*
> ```Go
> 
> 	// root is the root of the lowest cost expression tree in the memo. It is
> 	// set once via a call to the SetRoot method.
> ```

set once after optimization is complete

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-:-L9tux9Q-YoIGcBe832B:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/opt_index_selection.go, line 148 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-L9uBrHs2GqK0XOF2gCb:-L9uBrHs2GqK0XOF2gCc:b-7wyb5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a755e54db47204435255bdeedfcf835bbc5bb9cb/pkg/sql/opt_index_selection.go#L148)):*
> ```Go
> 			return nil, err
> 		}
> 		filterExpr := optimizer.Optimize(filterGroup, &memo.PhysicalProps{})
> ```

Why did you remove this optimization step? I guess for scalar expressions there is no exploration, only normalization?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-:-L9uDET8Bu2Q9186HH9d:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
*[pkg/sql/opt_index_selection.go, line 148 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-L9uBrHs2GqK0XOF2gCb:-L9uN0sD3v8bFlNXnVWl:b5vjmdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a755e54db47204435255bdeedfcf835bbc5bb9cb/pkg/sql/opt_index_selection.go#L148)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yeah, I just wanted to get an `ExprView` of the scalar expression and I think `Optimize` was the only available option at the time.
</blockquote></details>

Got it - thanks!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712)*
<!-- Sent from Reviewable.io -->

andy-kimball committed:
opt: Store root expression directly in Memo

Currently, the memo's root expression is maintained outside the memo.
There are a few cases (like printing the Memo) where having it
available in the memo would be convenient. This change sets the root
expression (defined as the root of the lowest cost expression tree)
just after optimization is complete. The memo formatting code now
uses that information directly. Future PR's will make further usage
of the root.

Release note: None

andy-kimball commented:
I made a number of changes to this PR as part of integrating with Justin's memo changes:

1. Added memo format flags, as I'm sure we'll continue adding formatting options in the future
2. Create `Root` accessor that returns an `ExprView`
3. Changed formatting order of root memo group so that root properties are always first

---

Review status: 2 of 17 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/opt/memo/memo.go, line 129 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-L9tuitw0bBiM4cWkbBi:-LA3v9p9-NjYEyOts-fX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7b586f08aeb888382ac0fa2579aa0cb86258d1f/pkg/sql/opt/memo/memo.go#L129)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

set once after optimization is complete
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-:-LA3vZkp69gSYtQQp-rF:bdqgaf9)*
<!-- Sent from Reviewable.io -->

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 17 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-:-LA9Z7S96maAomUWVWuF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 15 of 15 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/opt/memo/memo.go, line 170 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-LAEkUK59ZSHH3GzvOiW:-LAEkUK59ZSHH3GzvOiX:bw46m08) ([raw file](https://github.com/cockroachdb/cockroach/blob/39f55f81ae57212662b8e7b193644d1ff7a0934b/pkg/sql/opt/memo/memo.go#L170)):*
> ```Go
> 		panic("the given root is in a different memo")
> 	}
> 	if root.best == normBestOrdinal {
> ```

What if the best expression happens to be equivalent to the normalized tree? Does it still have a different ID?

---

*[pkg/sql/opt/memo/memo.go, line 384 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-LAEkKIyAeDuhpTMiMf4:-LAEkKIyAeDuhpTMiMf5:b-5d256h) ([raw file](https://github.com/cockroachdb/cockroach/blob/39f55f81ae57212662b8e7b193644d1ff7a0934b/pkg/sql/opt/memo/memo.go#L384)):*
> ```Go
> )
> 
> // String returns a human-readable string representation of this expression for
> ```

[nit] expression -> memo

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-:-LAEn9CSBxvXZz4s47Ke:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/opt/memo/memo.go, line 170 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-LAEkUK59ZSHH3GzvOiW:-LAF7RfX20FusWSmZs8x:btyecim) ([raw file](https://github.com/cockroachdb/cockroach/blob/39f55f81ae57212662b8e7b193644d1ff7a0934b/pkg/sql/opt/memo/memo.go#L170)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What if the best expression happens to be equivalent to the normalized tree? Does it still have a different ID?
</blockquote></details>

Yes, the normalized tree always has a different ID. The value of `normBestOrdinal` is `math.MaxUint16`, so if used as an index into the group's `BestExpr` values, it result in a panic.

---

*[pkg/sql/opt/memo/memo.go, line 384 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712#-LAEkKIyAeDuhpTMiMf4:-LAF806-EwR6OROvhEgc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/39f55f81ae57212662b8e7b193644d1ff7a0934b/pkg/sql/opt/memo/memo.go#L384)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] expression -> memo
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24712)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+
